cache_cleaner/cache_cleaner.py

```diff
@@ -17,68 +17,111 @@
+                stat = os.stat(filename, dir_fd=root_fd)
+                file_size = stat.st_size
+                last_access_time = stat.st_atime
+                age = now - last_access_time
+
+                # Delete any file that hasn't been accessed in > max_age
+                if age > max_age:
+                    delete_file(filename, root_fd)
+                    n_deletions += 1
+                    if not initial_pass:
+                        total_size -= file_size
+                    if not initial_pass and total_size <= max_cache_size:
+                        logger.info(
+                            "Cache reduced to %s, clearing complete",
+                            humanize.naturalsize(total_size, gnu=True)
+                        )
+                        return
+                    continue
+
+                # Store the least recently accessed file
+                if age > largest_access_age:
+                    largest_access_age = age
+
+                if initial_pass:
+                    total_size += file_size
+
+            for dirname in dirs:
+                delete_dir_if_empty(dirname, root_fd)
+
+        logger.info("Deleted %d files.", n_deletions)
+        initial_pass = False
+        if total_size > max_cache_size:
+            # Assuming that all files are the same size, how many would the
+            # excess size allow us to keep?
+            estimated_fraction_files_to_keep = max_cache_size / total_size
+            # Assume that access patterns follow a Pareto distribution
+            # ie something like "80% of accesses are for 20% of documents"
+            #
+            # We can approximate that the distribution of ages is similar
+            # to the distribution of document requests, where the latter
+            # is seen to be a Zipf distribution with a parameter of 0.8:
+            # http://seelab.ucsd.edu/mobile/related_papers/Zipf-like.pdf
+            #
+            # Following the relation between the Zipf and Pareto distributions,
+            # this makes our alpha for the Pareto distribution 1.8
+            # https://en.wikipedia.org/wiki/Pareto_distribution#Relation_to_Zipf's_law
+            #
+            # Thus, we can get the cutoff percentile of ages using the Lorenz
+            # curve for the Pareto distribution:
+            # https://en.wikipedia.org/wiki/Pareto_distribution#Lorenz_curve_and_Gini_coefficient
+            estimated_age_cutoff_percentile = 1 - pow(
+                1 - estimated_fraction_files_to_keep, 1 - (1 / 1.8)
+            )
+            # Next pass, start deleting files that are in this age percentile
+            max_age = largest_access_age * estimated_age_cutoff_percentile
+            logger.info(
+                "Cache size (%s) exceeds maximum size (%s), decreasing max age",
+                humanize.naturalsize(total_size, gnu=True),
+                humanize.naturalsize(max_cache_size, gnu=True)
+            )
```

Comment:
Minor: log the new max age? Would help us spot if there's a bug in this logic that reduces it to, say, 1 second before current time.

Reply:
It gets logged at the start of the next loop!

Reply:
Aha!